### PR TITLE
fix(main/neovim): turn nvim into a wrapper

### LINKED ISSUE #22328: [Bug]: Neovim cannot load native Lua modules that are not linked with LuaJIT
### Problem description

Neovim is unable to load lua modules created with [mlua](https://github.com/mlua-rs/mlua) if the module is not specifically linked with LuaJIT.
`require`-ing the module throws an error saying that dlopen cannot locate `lua_gettop` (or some other lua symbols).

This doesn't happen when the same code is run with `luajit` instead of `nvim`.

### What steps will reproduce the bug?

1. Install Neovim
2. Create a lua module with mlua (Like [this one](https://github.com/mlua-rs/mlua?tab=readme-ov-file#module-mode) but change the lua feature to `"luajit"`)
3. Build the module with `cargo build`
4. `cd` into `./target/debug`
5. `ln -s lib*.so my_module.so`
6. `nvim --clean --cmd "lua require('my_module')"`
7. See the error output

### What is the expected behavior?

The module is loaded and the function is accessible.

If the library is instead built with `RUSTFLAGS="-C link-args=-lluajit" cargo build` it works as expected, but this is not required on any other platforms and it makes cross compiling very difficult.

### System information

```shell
Termux Variables:
TERMUX_API_VERSION=0.50.1
TERMUX_APK_RELEASE=F_DROID
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=8828
TERMUX_IS_DEBUGGABLE_BUILD=0
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
TERMUX__USER_ID=0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://mirrors.tuna.tsinghua.edu.cn/termux/apt/termux-main stable main
# root-repo (sources.list.d/root.list)
deb https://mirrors.tuna.tsinghua.edu.cn/termux/apt/termux-root root stable
Updatable packages:
All packages up to date
termux-tools version:
1.44.3
Android version:
14
Kernel build information:
Linux localhost 5.15.104-android13-8-00010-g1dc49517c375-ab11006012 #1 SMP PREEMPT Wed Oct 25 06:11:20 UTC 2023 aarch64 Android
Device manufacturer:
Xiaomi
Device model:
23078RKD5C
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
Installed termux plugins:
com.termux.api versionCode:51
com.termux.widget versionCode:13
com.termux.styling versionCode:32
```
```

### PR DESCRIPTION
**Pull request by pipcet**
See #22328 for further discussion.

My main concern with this PR is that the workaround is likely brittle; Android's `dlopen`, which was already deliberately broken to prevent loading dynamic libraries, might very well be broken further by future Android versions when the Android people realize `LD_PRELOAD` is currently handled correctly.

Also fixes a minor idempotency issue (repeated builds of neovim failed previously, leaving `/data/data/com.termux/usr/bin/sh` as a symlink to `/bin/sh`).

### CODE REVIEW

**Comment by pipcet** [comment]
Sorry, forgot to bump the revision, making CI fail. Fixed now, I hope.

**Comment by pipcet** [comment]
@s-cerevisiae If you happen to have the time, it'd be great if you could confirm independently that this fixes the issue

**Comment by pipcet** [comment]
> @pipcet Sorry, it seems that it didn't work. Did you forget to export the variable?

You're totally correct. Both of my test systems had LD_PRELOAD already set and exported, so the `export` was indeed missing. Should work now?

**Comment by s-cerevisiae** [comment]
Ok I think it works for me this time. Thank you!

File: packages/neovim/build.sh [hunk 1]
```diff
@@ -82,7 +82,15 @@ termux_step_post_make_install() {
 	cp $TERMUX_PKG_BUILDER_DIR/sysinit.vim $_CONFIG_DIR/
 
 	# Tree-sitter grammars are packaged separately and installed into TERMUX_PREFIX/lib/tree_sitter.
-	ln -s "${TERMUX_PREFIX}"/lib/tree_sitter "${TERMUX_PREFIX}"/share/nvim/runtime/parser
+	ln -sf "${TERMUX_PREFIX}"/lib/tree_sitter "${TERMUX_PREFIX}"/share/nvim/runtime/parser
```
Comment (TomJo2000) [inline packages/neovim/build.sh:85]: Good catch on that, I noticed that was a problem earlier in testing but I hadn't gotten around to fixing it yet.

File: packages/neovim/build.sh [hunk 2]
```diff
@@ -82,7 +82,15 @@ termux_step_post_make_install() {
...
+
+	mv "${TERMUX_PREFIX}"/bin/nvim "${TERMUX_PREFIX}"/bin/nvim.bin
+	cat <<-EOF > "${TERMUX_PREFIX}"/bin/nvim
```
Comment (TomJo2000) [inline packages/neovim/build.sh:88]: I think this should be a separate file.
I'm also not the biggest fan of keeping the nvim binary in `/bin` if we're going to wrap it.

There's a couple other packages that get wrapped, I'd like to take a look at those first before making a final decision on if and how we should do the same for Neovim.

File: packages/neovim/build.sh [hunk 3]
```diff
@@ -82,7 +82,15 @@ termux_step_post_make_install() {
...
+#!/bin/sh
+export LD_PRELOAD="\${LD_PRELOAD}:libluajit.so"
+exec "${TERMUX_PREFIX}"/bin/nvim.bin "\$@"
+EOF
+	chmod a+x "${TERMUX_PREFIX}"/bin/nvim
```
Comment (TomJo2000) [inline packages/neovim/build.sh:93]:
```suggestion
	chmod u+x "${TERMUX_PREFIX}"/bin/nvim
```

File: packages/neovim/build.sh (hunk 2 above)
Comment (pipcet) [reply packages/neovim/build.sh:88]: Is there a good way to substitute `TERMUX_PREFIX` on install, though? That's why I chose the heredoc, but if there's a better way I'd gladly fix that.

libexec might be a better place for `nvim` (no need for the `.bin` suffix then)?

**Comment by TomJo2000** [comment]
There is another option I just remembered.
Though that might need additional testing.

Arch Linux loads an additional config file in the systemwide `sysinit.vim`.
https://gitlab.archlinux.org/archlinux/packaging/packages/neovim/-/blob/0.10.2-2/PKGBUILD#L93-101

In their case it's to add some RTP directories by default.
But we should be able to use the same approach to append to `LD_PRELOAD`,
without potentially running into issues with users using different `#!$PREFIX/bin/sh` implementations.

File: packages/neovim/build.sh (hunk 2 above)
Comment (TomJo2000) [reply packages/neovim/build.sh:88]: I just remembered another option entirely.
https://github.com/termux/termux-packages/pull/22343#issuecomment-2489183106

**Comment by pipcet** [comment]
> There is another option I just remembered. Though that might need additional testing.
> 
> Arch Linux loads an additional config file in the systemwide `sysinit.vim`. https://gitlab.archlinux.org/archlinux/packaging/packages/neovim/-/blob/0.10.2-2/PKGBUILD#L93-101
> 
> In their case it's to add some RTP directories by default. But we should be able to use the same approach to append to `LD_PRELOAD`, without potentially running into issues with users using different `#!$PREFIX/bin/sh` implementations.

I don't think I understand. `LD_PRELOAD` needs to be set before the `vim` executable is executed, or the dynamic loader won't notice it in time, IIUC.

**Comment by TomJo2000** [comment]
> I don't think I understand. `LD_PRELOAD` needs to be set before the `vim` executable is executed, or the dynamic loader won't notice it in time, IIUC.

Mmmh, yep LD_***PRE***LOAD, it's kind in the name...
So I guess `exec LD_PRELOAD="$LD_PRELOAD:libluajit.so" nvim "$@"` it is then.

**Comment by pipcet** [comment]
> > I don't think I understand. `LD_PRELOAD` needs to be set before the `vim` executable is executed, or the dynamic loader won't notice it in time, IIUC.
> 
> Mmmh, yep LD__**PRE**_LOAD, it's kind in the name... So I guess `exec LD_PRELOAD="$LD_PRELOAD:libluajit.so" nvim "$@"` it is then.

Just for completeness, one other way of working around this is to have `nvim` re-exec itself after modifying the environment, if necessary. I'm not sure it would be any better, though.

**Comment by TomJo2000** [comment]
Hmm the only really comparable wrapper I could dig up is the one for `tizonia`:
https://github.com/termux/termux-packages/blob/9b54716ef8c8bc63cc07060c22ab2273608fcc52/packages/tizonia/exe_wrapper#L1-L11
Although that serves a slightly different purpose than our needs here.
As it doesn't seem to be present in the final package and is only used as a shim in the build process.

**Comment by pipcet** [comment]
> If you can give me an example of how that would work, sure.
> 
> But I think a POSIX sh oneliner is probably the lesser evil here.

Does `neovim` load child processes? Because those would inherit the `LD_PRELOAD` setting, I believe, and that's quite undesirable.

**Comment by TomJo2000** [comment]
> Does `neovim` load child processes? Because those would inherit the `LD_PRELOAD` setting, I believe, and that's quite undesirable.

There's probably a way to make it do that.
But I'm not sure it's something we should be worried about.

**Comment by s-cerevisiae** [comment]
> Does neovim load child processes?

It does launch `nvim --embed` on startup but I'm afraid that's more of a implementation detail. And any config is probably processed in the `--embed` process. Imo using a simple wrapper script is more reliable than trying to trick the editor to do something.

**Comment by TomJo2000** [comment]
> And please don't do anything that can possibly slow down the startup time because, you know, it's nvim.

I think we'll be fine just eating the impact from appending and passing an environment variable and then calling `exec`.
You got a couple microseconds to spare.

**Comment by TomJo2000** [comment]
@pipcet I have pushed a branch to my `termux-packages` development fork with a slight variation on the approach you have taken in this PR currently.
https://github.com/TomJo2000/termux-packages-prs/commit/2df04d19155a186d27b0477df2a29ae26ea54ddd

With your permission I would like to push that changeset to this PRs branch so it can be merged as part of this PR.

That branch will also include your idempotency fix to the tree-sitter symlink, as well as some further housekeeping/general build script cleanup in a separate commit.

**Comment by pipcet** [comment]
> @pipcet I have pushed a branch to my `termux-packages` development fork with a slight variation on the approach you have taken in this PR currently. [TomJo2000@2df04d1](https://github.com/TomJo2000/termux-packages-prs/commit/2df04d19155a186d27b0477df2a29ae26ea54ddd) [TomJo2000@f6388bb](https://github.com/TomJo2000/termux-packages-prs/commit/f6388bb3eb42a98eb8250ed21a1a776ee574677f)
> 
> With your permission I would like to push that changeset to this PRs branch so it can be merged as part of this PR.

Absolutely, please go ahead and do that! The one problem I see is that `A=B exec env` works for me but `exec A=B env` doesn't.

**Comment by TomJo2000** [comment]
Just pushed the commits.

I can't make commits as you, since that would be identity theft.
<sup>Though since your commits aren't signed I could have changed the Author information to yours</sup>
So the best I can do is give you `Co-authored-by:` credits on both commits.

You may want to consider signing your commits if that's something you care about.
Your authentication SSH key can also be used as a signing key.
https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification

**Comment by pipcet** [comment]
> Just pushed the commits.

Thank you!

> I can't make commits as you, since that would be identity theft. Though since your commits aren't signed I could have changed the Author information to yours So the best I can do is give you `Co-authored-by:` credits on both commits.
> 
> You may want to consider signing your commits if that's something you care about. Your authentication SSH key can also be used as a signing key. https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification

Thanks for the advice. I've decided against signing commits for now, for various reasons, and I'm perfectly happy with the commits as you pushed them :-)

**Comment by robertkirkman** [comment]
> > Does neovim load child processes? Because those would inherit the `LD_PRELOAD` setting, I believe, and that's quite undesirable.

> It does launch `nvim --embed` on startup but I'm afraid that's more of a implementation detail. And any config is probably processed in the `--embed` process. Imo using a simple wrapper script is more reliable than trying to trick the editor to do something. (And please don't do anything that can possibly slow down the startup time because, you know, it's nvim.)

in theory, this is unfortunately a problem that can hypothetically cause issues in the future that are difficult to diagnose for someone who is not familiar with Termux, Neovim, Lua, the issue, and this PR.

**Comment by TomJo2000** [comment]
Damned if you do, damned if you don't.

- Since the new issue hasn't been reproducible by anyone other than the submitter so far,
and we're currently waiting on feedback regarding their environment specifics I think it's reasonable to keep the shim in place until we're able to get to the bottom of #30484

**Pull request merged**